digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala

```diff
@@ -1733,17 +1735,93 @@ class ProjectService(roadAddressService: RoadAddressService, roadLinkService: Ro
     }
   }
 
+  def handleTransferAndRenumeration(roadwayChanges: Seq[ProjectRoadwayChange]): Unit = {
+    var roadNames: Seq[RoadName] = Seq()
+    roadwayChanges.foreach(rwc => {
+      if (rwc.changeInfo.changeType.equals(AddressChangeType.ReNumeration) || rwc.changeInfo.changeType.equals(AddressChangeType.Transfer)) {
+        val targetRoadNumberOptional = rwc.changeInfo.target.roadNumber
+        val srcRoadNumberOptional = rwc.changeInfo.source.roadNumber
+
+        if (targetRoadNumberOptional.isDefined && srcRoadNumberOptional.isDefined) {
+          val targetRoadNumber = targetRoadNumberOptional.get
+          val srcRoadNumber = srcRoadNumberOptional.get
+
+          val targetExistingRoadways = roadwayDAO.fetchAllByRoadNumberAndValue(targetRoadNumber)
+          val srcExistingRoadways = roadwayDAO.fetchAllByRoadNumberAndValue(srcRoadNumber)
+
+          if (targetExistingRoadways.isEmpty) {
+            // TRANSFERING OR NUMBERING TO A NEW ROADWAY
+            if (srcExistingRoadways.nonEmpty) {
+              // IF EXISTS SOURCE ROADWAY AND ROADNAME EXPIRE IT
+              val srcRoadNames = RoadNameDAO.getAllByRoadNumber(srcRoadNumber)
+              if (srcRoadNames.nonEmpty) {
+                val roadNameOpt = srcRoadNames.find(rn => rn.endDate.isEmpty && rn.validTo.isEmpty)
+                if (roadNameOpt.isDefined) {
+                  val roadName = roadNameOpt.get
+                  RoadNameDAO.update(roadName.id, Map("endDate" -> rwc.projectStartDate.toLocalDate.toString("dd-MM-yyyy")))
+                }
+              }
+            }
+            // CREATE NEW ROADNAME FOR TARGET ROADNUMBER
+            val projectLinkNames = ProjectLinkNameDAO.get(Set(targetRoadNumber), rwc.projectId)
+            if (projectLinkNames.nonEmpty) {
+              roadNames = roadNames :+ RoadName(NewRoadNameId, targetRoadNumber, projectLinkNames.head.roadName, startDate = Some(rwc.projectStartDate), validFrom = Some(DateTime.now()), createdBy = rwc.user)
+              ProjectLinkNameDAO.removeProjectLinkName(targetRoadNumber, rwc.projectId)
+            }
+          } else {
+            // TRANSFERING OR NUMBERING TO AN EXISTING ROADWAY
+            // IF EXISTS SOURCE ROADWAY AND ROADNAME EXPIRE IT
+            val srcRoadNames = RoadNameDAO.getAllByRoadNumber(srcRoadNumber)
+            if (srcRoadNames.nonEmpty) {
+              val roadNameOpt = srcRoadNames.find(rn => rn.endDate.isEmpty && rn.validTo.isEmpty)
+              if (roadNameOpt.isDefined) {
+                val roadName = roadNameOpt.get
+                RoadNameDAO.update(roadName.id, Map("endDate" -> rwc.projectStartDate.toLocalDate.toString("dd-MM-yyyy")))
+              }
+            }
+          }
+        }
+      }
+    })
+
+    RoadNameDAO.create(roadNames)
+  }
+
+  def handleTerminatedRoadwayChanges(roadwayChanges: Seq[ProjectRoadwayChange]) = {
```

Comment:
In case, if only the whole road terminated, the road name can be terminated as well. If there are left other road parts of the same road, roan name remains valid

Reply:
I'm validating, by roadnumber, if there are any roadways. Only if there are no roadways, the road name is terminated